sbi/utils/metrics.py

```diff
@@ -20,22 +21,156 @@ def c2st(
+    two distributions <X> and <Y>. For details on the method, see [1,2].
+    If the returned accuracy is 0.5, <X> and <Y> are considered to be from the
+    same generating PDF, i.e. they can not be differentiated.
+    If the returned accuracy is around 1., <X> and <Y> are considered to be from
+    two different generating PDFs.
+
+    Training of the classifier with N-fold cross-validation [3] using sklearn.
+    By default, a `RandomForestClassifier` by from `sklearn.ensemble` is used
+    (<classifier> = 'rf'). Alternatively, a multi-layer perceptron is available
+    (<classifier> = 'mlp'). For a small study on the pros and cons for this
+    choice see [4].
+    If you need a more flexible interface which is able to take a sklearn
+    compatible classifier, see the `c2st_` method in this module.
+
+    Args:
+        X: Samples from one distribution.
+        Y: Samples from another distribution.
+        seed: Seed for the sklearn classifier and the KFold cross-validation
+        n_folds: Number of folds to use
+        z_score: Z-scoring using X, i.e. mean and std deviation of X is used to normalize Y using (Y - mean)/std
+        noise_scale: If passed, will add Gaussian noise with std noise_scale to samples of X and of Y
+        verbosity: control the verbosity of sklearn.model_selection.cross_val_score
+        classifier: classification architecture to use, possible values: 'rf' or 'mlp'
+
+    Return:
+        torch.tensor containing the mean accuracy score over the test sets
+        from cross-validation
+
+    Example:
+    ``` py
+    > c2st(X,Y)
+    [0.51904464] #X and Y likely come from the same PDF or ensemble
+    > c2st(P,Q)
+    [0.998456] #P and Q likely come from two different PDFs or ensembles
+    ```
+
+    References:
+        [1]: http://arxiv.org/abs/1610.06545
+        [2]: https://www.osti.gov/biblio/826696/
+        [3]: https://scikit-learn.org/stable/modules/cross_validation.html
+        [4]: https://github.com/psteinb/c2st/
+    """
+
+    clf_class = RandomForestClassifier
+    clf_kwargs = {}
+    if "mlp" in classifier.lower():
+        ndim = X.shape[-1]
+        clf_class = MLPClassifier
+        clf_kwargs = {
+            "activation": "relu",
+            "hidden_layer_sizes": (10 * ndim, 10 * ndim),
+            "max_iter": 1000,
+            "solver": "adam",
+            "early_stopping": True,
+            "n_iter_no_change": 50,
+        }
+    elif "rf" in classifier.lower():
+        # keep this section for later use
+        pass
+
+    scores_ = c2st_scores(
+        X,
+        Y,
+        seed=seed,
+        n_folds=n_folds,
+        scoring=scoring,
+        z_score=z_score,
+        noise_scale=noise_scale,
+        verbosity=verbosity,
+        clf_class=clf_class,
+        clf_kwargs=clf_kwargs,
+    )
+
+    scores = np.asarray(np.mean(scores_)).astype(np.float32)
+    value = torch.from_numpy(np.atleast_1d(scores))
+    return value
+
+
+def c2st_scores(
```

Comment:
I am confused, why do we need two functions here? the signatures are identical almost and the upper one is just delegating, no? 
One could outsource the classifier-building into a function like `build_classifier(classifier, kwargs)` but I see no need to have these two very similar functions. maybe I am missing something?

Reply:
yes, indeed. `c2st` is a quick interface and only returns the mean across the validation.
`c2st_scores` allows much more freedom and returns all scores from the cross validation.

Reply:
On a second thought, the signature of `c2st` can be reduced even more as `sbi` currently uses the defaults most exclusively in all cases.

Reply:
I am waiting for your suggestion on this one, but it feels to me that we are focusing very much on `c2st` here and the other metrics in this module remain as they are ... untested.

Reply:
I looked at this again and would vote for having only one `c2st` function and not `c2st` and `c2st_scores`. The cross validation belongs to c2st as it estimates the generalisation performance of the classifier--the key result we are interested in. So I suggest to join the two function, have the `classifier` string for passing `rf` or `mlp` (for now) and the `clf_kwargs` to update the default settings. We could add an additional `kwarg` that allows to return the `std` over the CV scores for analysis. 

Unless you have a strong opinion against this simplification?

Reply:
Ok, I did reduce the interface of `c2st` now with d456162
``` python
def c2st(
    X: Tensor,
    Y: Tensor,
    seed: int = 1,
    n_folds: int = 5,
    metric: str = "accuracy",
    classifier: str = "rf",
) -> Tensor:
```
versus
```
def c2st_scores(
    X: Tensor,
    Y: Tensor,
    seed: int = 1,
    n_folds: int = 5,
    metric: str = "accuracy",
    z_score: bool = True,
    noise_scale: Optional[float] = None,
    verbosity: int = 0,
    clf_class: Any = RandomForestClassifier,
    clf_kwargs: Dict[str, Any] = {},
) -> Tensor:
```
I personally don't see a merit in outsourcing building a classifyer in an independent function. But maybe I am getting something wrong here. Feel free to open an issue to suggest a refactoring in the future.